### ISSUE TITLE
Non-deterministic EnvironmentConfig choice with label selector

### ISSUE THREAD

**Issue by phisco** [crossplane/crossplane#4070]
I think the following should be addressed as part of #3770.

### What happened?

With the current implementation of EnvironmentConfigs, if a user specifies environments to be selected through a label selector instead of a direct reference, Crossplane lists all environment configs matching the given labels and picks the first one, see https://github.com/crossplane/crossplane/blob/234deec1a2657992ec078f04dd321cbf9a14e51f/internal/controller/apiextensions/composite/environment/selector.go#L125. This means that if the selector matches more than one EnvironmentConfig, depending on the ordering returned by the List call above which can not be assumed to be deterministic, a random one will be picked as the first one.

This is not a major issue, as given that selected environment configs are "stored" in the XR at `spec.environmentConfigRefs`, always the same configs are going to be used for a given XRC and therefore won't result in flapping values. But I think we should still enforce some ordering to ensure deterministic results across multiple Composite Resource or Claim creations.

#3931 is going to introduce the ability to match multiple EnvironmentConfigs and will result in randomly sorted environmentConfigs being stored in `spec.environmentConfigRefs` for each new XR or XRC created.

### How can we reproduce it?

Can be reproduced as follows:
```
# spin up a kind cluster
$ kind create cluster
...
# install crossplane with env configs enabled
$ helm upgrade --install crossplane --set args={"--enable-environment-configs"} crossplane-stable/crossplane
...
# deploy provider-nop just to have some dummy MR to use in a Composition
$ kubectl apply -f - <<EOF
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: crossplane-provider-nop
spec:
  package: crossplane/provider-nop:main
EOF
...
# wait for the provider to be ready
```
Deploy a simple XRD:
```
$ kubectl apply -f - <<EOF
---
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xsqlinstances.example.org
spec:
  group: example.org
  names:
    kind: XSQLInstance
    plural: xsqlinstances
  claimNames:
    kind: SQLInstance
    plural: sqlinstances
  defaultCompositionRef:
    name: nop.sqlinstances.example.org
  defaultCompositeDeletePolicy: Background
  defaultCompositionUpdatePolicy: Automatic
  versions:
  - name: v1alpha1
    served: true
    referenceable: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              parameters:
                type: object
                properties:
                  storageGB:
                    type: integer
                required:
                - storageGB
            required:
            - parameters
          status:
            type: object
            properties:
              address:
                description: Address of this MySQL server.
                type: string
EOF
...
```
Apply a few EnvironmentConfigs to be used by the Composition:
```
$ kubectl apply -f - <<EOF
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: EnvironmentConfig
metadata:
  name: example-environment
data:
  complex:
    a: b
    c:
      d: e
      f: "1"
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: EnvironmentConfig
metadata:
  name: example-environment-2
  labels:
    stage: prod
data:
  complex:
    c:
      f: "2" # <- same path, different value w.r.t. example-environment-3 below
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: EnvironmentConfig
metadata:
  name: example-environment-3
  labels:
    stage: prod
data:
  complex:
    c:
      f: "3" # <- same path, different value w.r.t. example-environment-2 above
EOF
...
```
Create a Composition with a label selector matching both example-environment-2 and example-environment-3, which have different values for `complex.c.f`, 2 and 3 respectively:
```
$ kubectl apply -f - <<EOF
---
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: nop.sqlinstances.example.org
spec:
  writeConnectionSecretsToNamespace: crossplane-system
  compositeTypeRef:
    apiVersion: example.org/v1alpha1
    kind: XSQLInstance
  environment:
    environmentConfigs:
    - type: Reference
      ref:
        name: example-environment
    - type: Selector
      selector:
        matchLabels:
        - type: FromCompositeFieldPath
          key: stage
          valueFromFieldPath: metadata.labels[stage]
  resources:
    - name: nop
      base:
        apiVersion: nop.crossplane.io/v1alpha1
        kind: NopResource
        spec:
          forProvider:
            conditionAfter:
            - time: 10s
              conditionType: Ready
              conditionStatus: "True"
          providerConfigRef:
            name: default
      patches:
        - type: FromEnvironmentFieldPath    # <- setting the value of complex.c.f from the computed environment, will be 2 or 3 according to whatever env is selected or would be the last selected if multiple envs are selected.
          fromFieldPath: complex.c.f
          toFieldPath: metadata.annotations[nop.crossplane.io/complex.c.f]
EOF
...
```
Now create multiple Claims and check the value set as annotation
```
$ i=0
while true; do
  cat <<EOF | kubectl apply -f -
---
apiVersion: example.org/v1alpha1
kind: SQLInstance
metadata:
  name: example-$i
  labels:
    stage: prod
spec:
  parameters:
    storageGB: 10
  writeConnectionSecretToRef:
    name: example-sqlinstance-connection
EOF
  echo "Press enter to continue, wait a few seconds for the claim to be created"
  read < /dev/tty
  kubectl get nopresources.nop.crossplane.io -l crossplane.io/claim-name=example-$i -o json | jq -r '.items[] | [.metadata.name,.metadata.annotations["nop.crossplane.io/complex.c.f"]]| @csv'
  i=$(($i+1))
done
```
would for example output the following:
```
sqlinstance.example.org/example-0 created
Press enter to continue, wait a few seconds for the claim to be created

"example-0-rlpsv-ztfcs","2"
sqlinstance.example.org/example-1 created
Press enter to continue, wait a few seconds for the claim to be created

...

"example-8-zgsxg-8zjlp","2"
sqlinstance.example.org/example-9 created
Press enter to continue, wait a few seconds for the claim to be created

"example-9-jw2kn-p5tc7","3"
sqlinstance.example.org/example-10 created
Press enter to continue, wait a few seconds for the claim to be created

"example-10-69bxr-rwkzj","2"
sqlinstance.example.org/example-11 created
Press enter to continue, wait a few seconds for the claim to be created

"example-11-b9cch-xjf7t","3"
sqlinstance.example.org/example-12 created
Press enter to continue, wait a few seconds for the claim to be created

"example-12-5g29l-4qlw9","2"
sqlinstance.example.org/example-13 created
Press enter to continue, wait a few seconds for the claim to be created

...
```

So as you can see above, the majority of resources are created with an environment specifying a value of `2` for `metadata.annotations["nop.crossplane.io/complex.c.f"]`, but some ends up with `3` being set. Which I don't think is a desirable inconsistency, so we should agree on a specific order to enforce:
- alphabetical by name: easy to think about, but not much flexible, users could prioritise environment configs by name. not the best.
- add a field to the `EnvironmentSourceSelector`, defaulting to `metadata.name` to specify the field configs should be sorted by.
- ???

Should be kept in mind that, if #3931 gets merged, it would reverse the sorting algorithm, as with the current approach the first element would win, while with that merged, the last one would.

### What environment did it happen in?
Crossplane version: 1.12.1

Upstream helm chart and image.

Not specific to Kubernetes versions and

**Comment by negz** [comment]
We actually lean into non-deterministic selection almost everywhere else in Crossplane. That is, if a selector matches multiple things we will actively make a random choice rather than picking the first item.

The idea is to:

* Motivate users to be specific with their labels. If it really matters that one thing is selected, make sure your labels can _only_ select that thing.
* Prevent users depending on quasi-determinism. For example if we list envconfigs and sort them by name a particular XR would always pick `env-config-desired` ...until someone else created `env-config-cool` for some entirely different purpose without realizing it will now supersede `env-config-desired`. This is a little like how Go [actively randomizes map iteration](https://nathanleclaire.com/blog/2014/04/27/a-surprising-feature-of-golang-that-colored-me-impressed/).

I believe this would be appropriate here as well - I would suggest we actively make a random selection rather than depending on the order of what is returned by the List call.

**Comment by phisco** [comment]
Thanks for clarifying this, @negz!

I do agree that selecting resources by label should not imply any particular order, that's what users expect as that's what they are used to, e.g. services selecting pods, operators discovering configs... Unfortunately here the order does matter as it could potentially change the end result of a patch. Actively randomising the order would lead to flapping values if the Resolve Policy introduced by https://github.com/crossplane/crossplane/pull/3931 was set to Always, as you can see [here](https://github.com/crossplane/crossplane/pull/3981/files#diff-eb972d433230cf1b5f44ddc651e78cb0f149578d1eae24b4da77093efd605bccR82).

**Comment by negz** [comment]
> Actively randomising the order would lead to flapping values if the Resolve Policy introduced by https://github.com/crossplane/crossplane/pull/3931 was set to Always, as you can see [here](https://github.com/crossplane/crossplane/pull/3981/files#diff-eb972d433230cf1b5f44ddc651e78cb0f149578d1eae24b4da77093efd605bccR82).

Ah, I did not realize you were talking about this proposed design that selects _multiple_ EnvConfigs rather than selecting one. Based on my understanding of the design I agree those need to be sorted, not random.

In cases where we select _one_ thing I do think we should randomize instead of giving the false impression of "stable" selection.

**Comment by MisterMX** [comment]
I personally would advise to NOT select multiple configs at once as the resulting data block depends highly on the state of the cluster and the number of existing configs. I cannot think of a good use case where you would want to select an arbitrary number of unknown configs with an unknown structure and merge them in an order that you don't know.

I would go so far and even vote for returning an error if multiple configs are the result of a selection since we are not able to determine which one to actually use and, as @negz pointed out, Crossplane almost always makes a random selection - which doesn't make sense here imo.

**Comment by phisco** [comment]
Totally agree the random selection doesn't make sense here, neither if we keep picking a single one, nor with the change implemented in #3981. In the former case, I do agree returning an error would be the best thing, although it would be yet another confusing runtime error, which could potentially be triggered by an external user either by mistake or willingly creating an additional EnvironmentConfig.

**Comment by phisco** [comment]
Solved by sorting by a user defined field path by: https://github.com/crossplane/crossplane/pull/3981
